packager/file/http_file.cc

```diff
@@ -333,8 +334,39 @@ void HttpFile::SetupRequest() {
 
 void HttpFile::ThreadMain() {
   SetupRequest();
+  CURLcode res;
+  int64_t sleep_duration = kFirstRetryDelayMilliseconds;
+
+  // Try sending the request multiple times if it fails.
+  for (int i = 0; i < kNumRequestErrorRetries; ++i) {
+    res = curl_easy_perform(curl_.get());
+
+    if (res == CURLE_HTTP_RETURNED_ERROR) {
+      long res_code = 0;
+      curl_easy_getinfo(curl_.get(), CURLINFO_RESPONSE_CODE, &res_code);
+      if (res_code == 408 || res_code == 429 || res_code >= 500) {
+        // FIXME: Should we clear download_cache_ and upload_cache_ here??
```

Comment:
Yes, you'll need to ensure that the inside of the loop has no side-effects.  I haven't analyzed to see what the caches do here, but it's worth looking into more closely.

You may want to write test cases that verify the correct behavior for POST/PUT with data.  In the `cmake` branch, you'll find HTTP file tests are all enabled and working.  (In `main`, they are all disabled.)  You should be able to verify retry behavior with httpbin, and see that the final failed retry still has a response that reflects the data from the POST/PUT.

I recommend you try the `cmake` branch in a separate clone, without using gclient.  We have yet to update the docs for that, but this is what you'll want to try:

```sh
# Clone
git clone ...
# Use cmake branch
git checkout -b cmake upstream/cmake
# Get submodules
git submodule init && git submodule update
# Gen build files
cmake -S . -B build/ -DCMAKE_BUILD_TYPE=Debug
# Build
cmake --build build/ --config Debug --parallel
# Test
(cd build && ctest -C Debug -V)
```

If you have time, you could make a PR against the `cmake` branch, which is where most development is happening right now as we rewrite the build system and dependencies.